### PR TITLE
Refactor the templates used to build the UCCSD ansatz

### PR DESCRIPTION
**Pull request by agran2018**
**Context:**
This is part of the PL-QChem refactoring.

**Description of the Change:**
The functions implementing the quantum circuits building the Unitary Coupled-Cluster
(UCCSD) VQE ansatz have been improved, with a more consistent naming convention and
improved docstrings.

The changes include:

  - The terms *1particle-1hole (ph)* and *2particle-2hole (pphh)* excitations
    were replaced with the names *single* and *double* excitations, respectively.

  - The non-differentiable arguments in the `UCCSD` template were renamed accordingly:
    `ph` → `s_wires`, `pphh` → `d_wires`

  - The term *virtual*, previously used to refer the *unoccupied* orbitals, was discarded.

  - The Usage Details sections were updated and improved.

**Benefits:**
The new naming convention is more consistent. This makes the connection with the relevant functions in the `structure` module more evident. Furthermore, the argument descriptions have been improved and the Usage Details sections have been updated accordingly.

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: pennylane/templates/subroutines/double_excitation_unitary.py [hunk 1]
```diff
@@ -471,13 +474,12 @@ def DoubleExcitationUnitary(weight, wires1=None, wires2=None):
             dev = qml.device('default.qubit', wires=5)
 
             @qml.qnode(dev)
-            def circuit(weight, pphh=None):
```
Comment (josh146) [inline pennylane/templates/subroutines/double_excitation_unitary.py:474]: Was this example previously out of date with the docstring?

File: pennylane/templates/subroutines/uccsd.py [hunk 2]
```diff
@@ -68,30 +68,29 @@ def UCCSD(weights, wires, ph=None, pphh=None, init_state=None):
...
-        pphh (Sequence[Sequence[Sequence]]): Sequence of lists, each containing two lists that specify the indices
-            ``[s, ...,r]`` and ``[q,..., p]`` defining the 2particle-2hole configurations (pphh)
-            :math:`\vert \mathrm{pphh} \rangle = \hat{c}_p^\dagger \hat{c}_q^\dagger \hat{c}_r
-            \hat{c}_s \vert \mathrm{HF} \rangle`. The entries ``s`` and ``r`` are wires representing
-            two occupied orbitals where the two particles are annihilated (holes created)
-            while the entries ``q`` and ``p`` correspond to the wires representing two virtual
-            orbitals where the particles are created. Wires in-between represent the occupied
-            and virtual orbitals in the intervals ``[s, r]`` and ``[q, p]``, respectively.
+            :func:`~.DoubleExcitationUnitary`. These parameters are the coupled-cluster
+            amplitudes that need to be optimized for each single and double exciation generated
```
Comment (josh146) [inline pennylane/templates/subroutines/uccsd.py:76]:
```suggestion
            amplitudes that need to be optimized for each single and double excitation generated
```

File: pennylane/templates/subroutines/uccsd.py [hunk 3]
```diff
@@ -118,31 +117,34 @@ def UCCSD(weights, wires, ph=None, pphh=None, init_state=None):
...
-                UCCSD(weights, wires, ph=ph, pphh=pphh, init_state=ref_state)
-                return qml.expval(qml.PauliZ(0))
+            # Define the UCCSD ansatz
+            ansatz = partial(UCCSD, init_state=ref_state, s_wires=s_wires, d_wires=d_wires)
 
-            params = np.random.normal(0, np.pi, len(ph) + len(pphh))
-            print(circuit(params, wires, ph=ph, pphh=pphh, init_state=ref_state))
+            # Compute the expectation value of 'h' for given set of parameters 'params'
+            params = np.random.normal(0, np.pi, len(singles) + len(doubles))
+            print(qml.VQECost(ansatz, h, dev))
```
Comment (josh146) [inline pennylane/templates/subroutines/uccsd.py:147]: Just checking that you have double checked that this example still runs after the changes!

File: pennylane/templates/subroutines/double_excitation_unitary.py (hunk 1 above)
Comment (agran2018) [reply pennylane/templates/subroutines/double_excitation_unitary.py:474]: Yes, when we did the generalization of the templates we missed this :-(.

File: pennylane/templates/subroutines/uccsd.py (hunk 3 above)
Comment (agran2018) [reply pennylane/templates/subroutines/uccsd.py:147]: Thanks @josh146. In order to check it within this PR I would need to use the old names of the functions that were refactored but have not been merged yet. I can go with the old functions, check and update the names afterwards or update the name of the functions here and double check before merging. I will keep it in mind.

File: pennylane/templates/subroutines/double_excitation_unitary.py [hunk 4]
```diff
@@ -456,10 +459,10 @@ def DoubleExcitationUnitary(weight, wires1=None, wires2=None):
                H, R_x(-\frac{\pi}{2}), R_z(-\frac{\theta}{8}) \bigg] \Bigg\}
 
         #. For a given quadruple ``[s, r, q, p]`` with :math:`p>q>r>s`, seventy-two single-qubit
-           operations are applied. Notice also that consecutive CNOT gates act on qubits with
-           indices between ``s`` and ``r`` and ``q`` and ``p`` while a single CNOT acts on wires
-           ``r`` and ``q``. The operations performed across these qubits are shown in dashed lines
-           in the figure above.
+           and ``16*(len(wires1)-1 + len(wires2)-1 + 1)`` CNOT operations are applied. Notice also
+           that consecutive CNOT gates act on qubits with indices between ``s`` and ``r`` and
```
Comment (ixfoduap) [inline pennylane/templates/subroutines/double_excitation_unitary.py:463]:
```suggestion
           and ``16*(len(wires1)-1 + len(wires2)-1 + 1)`` CNOT operations are applied. Consecutive CNOT gates act on 
           qubits with indices between ``s`` and ``r`` and
```

File: pennylane/templates/subroutines/single_excitation_unitary.py [hunk 5]
```diff
@@ -49,32 +49,39 @@ def SingleExcitationUnitary(weight, wires=None):
...
     the exponential of a tensor product of Pauli-Z operators can be decomposed in terms of
-    :math:`2(n-1)` CNOT gates and a single-qubit Z-rotation. If there are :math:`X` or
-    :math:`Y` Pauli matrices in the product, the Hadamard (:math:`H`) or :math:`R_x` gate has
-    to be applied to change to the :math:`X` or :math:`Y` basis, respectively.
+    :math:`2(n-1)` CNOT gates and a single-qubit Z-rotation referred to as :math:`U_\theta` in
+    the figure above. If there are :math:`X` or :math:`Y` Pauli matrices in the product,
+    the Hadamard (:math:`H`) or :math:`R_x` gate has to be applied to change to the
+    :math:`X` or :math:`Y` basis, respectively. The latter operations are denoted as
+    :math:`U_1` and :math:`U_2` in the figure above. See the Usage Details section for more
+    details.
```
Comment (ixfoduap) [inline pennylane/templates/subroutines/single_excitation_unitary.py:76]:
```suggestion
    information.
```

**Pull request merged**